com.microsoft.mrtk.input/Visualizers/ControllerVisualizer/ControllerModelLoader.cs

```diff
@@ -23,53 +26,92 @@ public static class ControllerModelLoader
         private static Dictionary<ulong, GameObject> controllerModelDictionary = new Dictionary<ulong, GameObject>();
 
         /// <summary>
-        /// Tries to load the controller model game object from the provided OpenXR ControllerModel.
+        /// A dictionary which caches the controller model gameobject associated with a specified input device
+        /// Stores a boolean indicating whether a warning was recently raised when trying to get the model key for this input device.
+        /// </summary>
+        private static Dictionary<InputDevice, bool> warningCache = new Dictionary<InputDevice, bool>();
+
+        /// <summary>
+        /// A dictionary which caches the controller model gameobject associated with a specified input device
+        /// Stores a boolean indicating whether an error was recently raised when trying to get the model for this model key.
+        /// </summary>
+        private static Dictionary<ulong, bool> errorCache = new Dictionary<ulong, bool>();
+
+        /// <summary>
+        /// Tries to load the controller model game object of the specified input device with the corresponding handedness
         /// Requires the MR OpenXR plugin to work.
         /// </summary>
-        /// <param name="controllerModelProvider">The OpenXR ControllerModel to load from.</param>
-        /// <returns>A GameObject representing the generated controller model in the scene.</returns>
-        public async static Task<GameObject> TryGenerateControllerModelFromPlatformSDK(Handedness handedness)
+        /// <param name="inputDevice">The input device we are trying to get the controller model of</param>
+        /// <param name="handedness">The handedness of the input device requesting the controller model</param>
+        /// <returns>A gameobject representing the generated controller model in the scene</returns>
+        public async static Task<GameObject> TryGenerateControllerModelFromPlatformSDK(InputDevice inputDevice, Handedness handedness)
```

Comment:
Conceptually, I'm trying to think through the potential long-term impacts of updating this API to have a strong dependency on Unity's input system types, when they're only used for caching.

_Hopefully_, we'll be on this input system for a while and Unity won't change things up again any time soon, but I remember the holes we dug ourselves in both HTK and MRTK2 when Unity updated their input methods, types, and patterns.

Mostly just wanted to type all this out to think this through and make sure we don't see a world where we'll want to support this API but _not_ key off of the Unity input system for attempted load.

Reply:
The inclusion of the inputDevice parameter is mostly just there to surpress a warning per-attempted input device load. Keying off of just handedness I think is _technically_ incomplete as far as an implementation goes, and opens us up to scenarios where don't show a warning where it could technically be warranted (i.e. one controller has a slightly different set of capabilities but doesn't support loading a controller model, which gets switched with a different controller which is able to load a model). Again, it's pretty edge case stuff, since users technically shouldn't be hot swapping controllers on the same hand constantly, but that's mostly because we don't think of using a MR application for days at a time currently, like we might a web-browser instance.